[AOTI] Add a boxed_run API

desertfire opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #142213

Summary: Fixes https://github.com/pytorch/pytorch/issues/141696. Add a new C++ runner API (boxed_run) following dynamo's boxed calling convention, which steals tensors' ownership from the input tensor list.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @chauhang @aakhundov @BoyuanFeng

desertfire committed:
Update

[ghstack-poisoned]

desertfire committed:
Update

[ghstack-poisoned]

desertfire commented:
This PR will conflict with https://github.com/pytorch/pytorch/pull/140664 .  I will only merge this one after  https://github.com/pytorch/pytorch/pull/140664 lands.

ezyang commented on torch/csrc/inductor/aoti_runner/model_container_runner_cuda.cpp:
```diff
@@ -23,6 +23,17 @@ std::vector<at::Tensor> AOTIModelContainerRunnerCuda::run(
 }
 
+std::vector<at::Tensor> AOTIModelContainerRunnerCuda::steal_inputs_and_run(
+    std::vector<at::Tensor>& inputs,
```
Have you actually checked this steals? E.g., test that the list is empty in Python at the end. Based on my understanding of how pybind11 is implemented, it's hard for me to imagine this actually works. (I think you will need to take a py::list directly and clear it by hand after you coerce the tensors out).

ezyang reviewed (changes requested):
Needs tests

ezyang commented:
You might consider calling it "boxed run" because that's the prevailing naming convention for this calling convention

desertfire committed:
Update

[ghstack-poisoned]

desertfire committed:
Update

[ghstack-poisoned]

ezyang commented on test/inductor/test_aot_inductor.py:
```diff
@@ -4100,6 +4102,59 @@ def forward(self, x):
+
+        model = Model().to(device=self.device)
+        # NOTE: There are additional references to inps if we use
+        # strict=True here, which will cause inps not deallocated
```
cc @albanD could this be reference cycles?

albanD replied:
Does gc.collect() after exiting this function makes the inps go away? Or you still hold onto that memory?

But in general I'm quite confused by this test, how is len(inps) == 0 ever ?!
Also inp_refs_0 is never checked? So it can be removed? (looking forward to the ruff rule for unused variable to be enabled to catch these :D )

desertfire replied:
This test is modified from the one in `test_torchinductor`. I tried to call `gc.collect()` in the later `__torch_dispatch__ ` function, similar to how `test_torchinductor.test_list_clearing` does it, but it didn't help. I checked the referrers to `inps`, and there are two extra referrers which look like some closures after running strict-mode export (i.e. dynamo), but it's not obvious to me where those extra references are coming from. cc @anijain2305 in case you have any ideas.

`inp_refs_0` was also modified based on the original test which I didn't end up using (explained in later comments). I can remove it if it is distracting.

albanD replied:
Running collect() within the function most likely won't change anything indeed. you need to run it after exiting to make sure all local variables are gone. 
You can wrap your whole test in a helper for that if needed.

Note that I am ALSO confused by the test in `test_torchinductor`: how does len(inps) become 0 ??

desertfire replied:
Calling `gc.collect()` after the exiting doesn't help to achieve what this test is targeting.

`len(inps)` becomes 0 because inductor-generated code explicitly clears input arg list as a part of the boxed run convention.

ezyang commented on test/inductor/test_aot_inductor.py:
```diff
@@ -4100,6 +4102,59 @@ def forward(self, x):
+                    # refs here invalid.
+                    # test_self.assertIsNone(inp_refs[0]())
+                    # test_self.assertIsNone(inp_refs[1]())
+                return func(*args, **kwargs)
```
Using a TorchDispatchMode to interpose on the torch calls inside the compiled code feels... fragile? Maybe AOTI has better guarantees about this because the runtime interface is well defined.

I'm trying think of a better, extensional way of verifying the inputs get deallocated in a timely fashion. Another idea is to put a custom deleter on the tensors, and then show that those tensors are deleted before some custom operator is called in AOTI.

desertfire replied:
One way I can think of is to introduce a C shim function just for debugging purpose, and introduce a special flag that codegening calls to that function to check if certain tensors are deleted. Let me know how that sounds to you.

ezyang replied:
I'm vaguely not positive on adding another codegen mode for debugging, mostly because that's more codegen code to have to maintain. Is there something you don't like about my idea?

desertfire replied:
> then show that those tensors are deleted before some custom operator is called in AOTI

If this is the idea you are referring to, I am fine with it and I think we are along the same line. When I said add a C shim function, I was referring to your comment about "some custom operator".

ezyang replied:
oh ok, I was just not expecting to need to add something new to the C API, boxed call for arbitrary custom ops should work right? :)

ezyang commented on torch/csrc/inductor/aoti_package/pybind.cpp:
```diff
@@ -5,26 +5,64 @@
+    inputs.clear();
+
+    std::vector<at::Tensor> result_tensors =
+        AOTIModelPackageLoader::boxed_run(input_tensors, stream_handle);
```
You should be `std::move(input_tensors)` into boxed run, that gives the right semantics. (Which means boxed_run should take the vector as rvalue ref.)

desertfire replied:
Will update.

desertfire commented on torch/csrc/inductor/aoti_runner/model_container_runner_xpu.h:
```diff
@@ -23,13 +23,13 @@ class C10_EXPORT AOTIModelContainerRunnerXpu : public AOTIModelContainerRunner {
-      const std::vector<at::Tensor>& inputs,
-      void* stream_handle = nullptr) override;
+  std::vector<at::Tensor> run_impl(
+      std::vector<AtenTensorHandle>& input_handles,
```
There is no real signature change on `run`. `AOTIModelContainerRunnerXpu` as a child class only overrides `run_impl`. `run`'s signature stays the same, but I have removed it from this header file since we are not overriding it.

desertfire committed:
Update

[ghstack-poisoned]

desertfire committed:
Update on "[AOTI] Add a boxed_run API"


Summary: Fixes https://github.com/pytorch/pytorch/issues/141696. Add a new C++ runner API (boxed_run) following dynamo's boxed calling convention, which steals tensors' ownership from the input tensor list.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames chauhang aakhundov

[ghstack-poisoned]

desertfire committed:
Update on "[AOTI] Add a boxed_run API"


Summary: Fixes https://github.com/pytorch/pytorch/issues/141696. Add a new C++ runner API (boxed_run) following dynamo's boxed calling convention, which steals tensors' ownership from the input tensor list.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames chauhang aakhundov

[ghstack-poisoned]

desertfire committed:
Update on "[AOTI] Add a boxed_run API"


Summary: Fixes https://github.com/pytorch/pytorch/issues/141696. Add a new C++ runner API (boxed_run) following dynamo's boxed calling convention, which steals tensors' ownership from the input tensor list.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames chauhang aakhundov

[ghstack-poisoned]

desertfire committed:
Update on "[AOTI] Add a boxed_run API"


Summary: Fixes https://github.com/pytorch/pytorch/issues/141696. Add a new C++ runner API (boxed_run) following dynamo's boxed calling convention, which steals tensors' ownership from the input tensor list.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames chauhang aakhundov

[ghstack-poisoned]

desertfire commented on benchmarks/dynamo/expected_ci_speedup_inductor_torchbench_cpu.csv:
```diff
@@ -12,7 +12,7 @@ functorch_maml_omniglot,inductor,float32,dynamic,cpp,1.126799
 mobilenet_v2,export-aot-inductor,float32,static,default,2.90375357
 resnext50_32x4d,export-aot-inductor,float32,dynamic,default,1.49299689
-hf_Albert,export-aot-inductor,float32,dynamic,default,1.33293645
+hf_Albert,export-aot-inductor,float32,dynamic,default,1.261471
```
I was not able to reproduce the slowdown locally, https://github.com/pytorch/pytorch/actions/runs/12586340292/job/35083670600. Updating the threshold for now.

desertfire commented:
@pytorchbot merge

kit1980 commented:
evert -m "breaking lots of internal builds, see D68036023" -c ghfirst

desertfire committed:
Update

[ghstack-poisoned]

desertfire commented:
@pytorchbot merge
